quipucords/api/credential/serializer.py

```diff
@@ -244,3 +244,80 @@ def prepare_data_for_update(cls, data: dict):
             data["ssh_keyfile"] = None
         elif ssh_keyfile:
             data["password"] = None
+
+
+class AuthTokenOrUserPassSerializer(CredentialSerializer):
+    """Serialize credentials that require username+password or auth token."""
+
+    class Meta:
+        """Serializer configuration."""
+
+        model = Credential
+        fields = [
+            "cred_type",
+            "id",
+            "name",
+            "password",
+            "username",
+            "auth_token",
+            "sources",
+        ]
+        extra_kwargs = {
+            "password": {**ENCRYPTED_FIELD_KWARGS},
+            "auth_token": {**ENCRYPTED_FIELD_KWARGS},
+        }
+
+    def validate(self, attrs: dict):
+        """Run validations that require more than one field."""
+        attrs = super().validate(attrs)
+        has_user_or_pass, has_auth_token = self._get_user_pass_or_auth_token(attrs)
+
+        if not has_auth_token and not has_user_or_pass:
+            raise ValidationError(_(messages.TOKEN_OR_USER_PASS))
+
+        if has_auth_token and has_user_or_pass:
+            raise ValidationError(_(messages.TOKEN_OR_USER_PASS_NOT_BOTH))
+        # defer the rest of the validation to specialized serializers
+        if has_auth_token:
+            none_fields = {"username": None, "password": None}
+            validator_class = AuthTokenSerializer
+        else:
+            none_fields = {"auth_token": None}
+            validator_class = UsernamePasswordSerializer
+        self._setup_default_data(attrs)
+        # force values to some fields to None to ensure non-sense combinations (like
+        # auth_token + username) can't happen
+        attrs.update(**none_fields)
```

Comment:
Could this ever actually happen? Rather than silently squash potentially bad data, should we treat bad data as an error and raise a validation exception?

Reply:
>Could this "nonsense combination" ever actually happen? 

It will happen if you switch from "auth token" to "username+password". 2 of the new tests are actually covering that.

>Rather than silently squash potentially bad data, should we treat bad data as a 400 error and raise a validation exception?

I usually prefer throwing errors with non-sense data, but for this case I feel the UX would be bad (we would need to require that username + password is explicitly set to None when switching to auth_token.

Reply:
Overriding this data with null is equivalent to this:
https://github.com/quipucords/quipucords/blob/486a5a1245d1387516acd79e3745ca3536afc7b0/quipucords/api/credential/serializer.py#L236-L246

Reply:
ack. Thanks for the explanation!